### PR TITLE
Push rocm to slow path

### PR DESCRIPTION
**Pull request by izdeby**
Push rocm to slow path

### CODE REVIEW

**Comment by ngimel** [comment]
cc @jeffdaily. foreach operations (operations on the lists of tensors) are also passing a struct by value as a kernel argument, so given rocm poor perf with this pattern we are disabling it.

**Review by ngimel — APPROVED** [review]
Approving, let's monitor how long ROCm CI takes (hopefully should be manageable)

**Comment by jeffdaily** [comment]
Can we have some perf results to back this up before we make a change like this?

**Comment by izdeby** [comment]
> Can we have some perf results to back this up before we make a change like this?

@jeffdaily, what perf results would you like to see exactly? We will see rocm build time on CI once this lands.

**Comment by jeffdaily** [comment]
Are there any benchmarks like the ones used to verify #46097?

**Comment by ngimel** [comment]
@jeffdaily  you can use something like this to benchmark
```
import torch
import time
import torch.optim as optim
from torch.autograd import Variable
from torch.optim.lr_scheduler import ExponentialLR, ReduceLROnPlateau, StepLR
import torch.nn as nn
import time
import torchvision
import torch.utils.benchmark as benchmark_utils

device = "cuda"
model = torchvision.models.resnet.resnet101(pretrained=True).to(device)
targets = torch.randint(0, 1000, (100, 100), device=device)
criterion = nn.CrossEntropyLoss()

optimizer = optim.SGD(model.parameters(), lr=1e-3, momentum=0.1) # <----------------------- optimizer.
                                                          # would compare optim.SGD vs optim._multi_tensor.SGD
optimizer_mta = optim._multi_tensor.SGD(model.parameters(), lr=1e-3, momentum=0.1)
running_loss = 0.0
target = torch.empty(128, dtype=torch.long, device=device).random_(5)

optimizer.zero_grad()
inputs = torch.rand(128, 3, 100, 100, device=device , requires_grad=True)
outputs = model(inputs)
loss = criterion(outputs, target)
loss.backward()
optimizer.step()
running_loss += loss.item()

def main():
    timer = benchmark_utils.Timer(
        stmt="torch.cuda.synchronize();optimizer.step()",
        globals=globals(),
        label="str(optimizer)",
    )

    timer_mta = benchmark_utils.Timer(
        stmt="torch.cuda.synchronize(); optimizer_mta.step()",
        globals=globals(),
        label="str(optimizer_mta)",
    )
    for _ in range(1):
        for i in range(1):
           print(f"Run: {i}\n{'-' * 40}")
    #       print(f"timeit:\n{timer.timeit(1000)}\n")
           print(f"autorange:\n{timer.blocked_autorange()}\n\n")

        for i in range(1):
            print(f"Run: {i}\n{'-' * 40}")
    #        print(f"timeit:\n{timer_mta.timeit(1000)}\n")
            print(f"autorange:\n{timer_mta.blocked_autorange()}\n\n")

if __name__ == "__main__":
    main()
```

**Comment by jeffdaily** [comment]
@izdeby so we're really going to land this PR without any performance results to substantiate the need? CC @sunway513

**Comment by izdeby** [comment]
@jeffdaily, at this point of time i just merged the latest master. Few things, im not sure what benchmarks are we looking for here as im pretty much soft disabling rocm for foreach APIs due to existing perf issues on rocm side. Also, i dont have a rocm machine to collect any benchmarks in a first place.

**Comment by jeffdaily** [comment]
Results of @ngimel 's benchmark.
Before:
```
Run: 0
----------------------------------------
autorange:
<torch.utils.benchmark.utils.common.Measurement object at 0x7f5ea384b550>
str(optimizer)
  Median: 17.15 ms
  2 measurements, 10 runs per measurement, 1 thread


Run: 0
----------------------------------------
autorange:
<torch.utils.benchmark.utils.common.Measurement object at 0x7f5ea384b550>
str(optimizer_mta)
  207.44 ms
  1 measurement, 10 runs , 1 thread
```
After
```
Run: 0
----------------------------------------
autorange:
<torch.utils.benchmark.utils.common.Measurement object at 0x7fe904349470>
str(optimizer)
  Median: 17.48 ms
  2 measurements, 10 runs per measurement, 1 thread


Run: 0
----------------------------------------
autorange:
<torch.utils.benchmark.utils.common.Measurement object at 0x7fe904349470>
str(optimizer_mta)
  Median: 13.86 ms
  2 measurements, 10 runs per measurement, 1 thread
```

**Comment by jeffdaily** [comment]
@ngimel is there timeline when the optimizers based on multi_tensor_apply will become the default for pytorch?

**Comment by ngimel** [comment]
cc @izdeby. As soon as we can support all optimizers with type promotion, should be a few months. 
Note, sending ROCm to "slow" path for multi-tensor apply operations does not make it meaningfully slower compared to what happens in optimizers now - the operation is called in a loop for tensors in a list in both cases. For now ROCm just cannot take advantage of the speed-ups that multi-tensor-apply provides on cuda (similar thing happened with `cat` that you guys worked around).
